### PR TITLE
Betters deps for ocamlopt-memtrace

### PR DESCRIPTION
**Pull request by Gbury**

### CODE REVIEW

**Comment by lthls** [comment]
I doubt it fixes the bug.
The problem is not that these libraries are not yet up to date when compiling memtrace, the problem is that `make` will not consider that the `ocamlopt-memtrace.opt` binary depends on them, so even if they're more recent it will not trigger a recompilation.
In fact, if you look at the command used to generate the binary, it passes `unix.cmxa threads.cmxa ocamlcommon.cmxa ocamloptcomp.cmxa` on the command line without depending on them.
There's also the small issue of the memtrace `main.cm*` not depending on `driver/optmaindriver.cm*`, which can be fixed by setting the `depincludes` variable correctly.

If you want to test that a fix works, do a clean checkout then `make world.opt`, then patch `optmaindriver.ml` to print something at toplevel, and run `make world.opt` again. `ocamlopt.opt` should print the extra line, and if `ocamlopt-memtrace.opt` does not, the problem is not fixed.

**Comment by Gbury** [comment]
I agree with the missing dep on `driver/optmaindriver.ml` and will patch that.
For the rest of the deps, as far as my understanding of makefile goes, dummy targets (which have no dependencies), are supposed to be considered "fresh" every time and thus re-run each time make is called, so as it stands with this PR, `make ocamlopt-memtrace.opt` should always trigger a recompilation of otherlibrariesopt and stdlib (well, rather the corresponding makefile invocation), which should make them up-to-date but also means that the `ocamlopt-memtrace.opt` binary should always be recompiled (or rather, the commands associated to the target always run, even if it doesn't need it). Is that not correct ?

**Comment by Gbury** [comment]
Also, I don't know if it's possible for the `ocamlopt-memtrace.opt` target in `memprof/memtrace/Makefile` to depend on the `*.cmxa` since they are not within (or below) the directory where the makefile is; when i tried to add them as deps, makefile complained it didn't know how to build them, so I only fixed the deps in the root makefile, but if you have a solution, that'd be nice, ^^
EDIT: I think I see the problem of not recompiling.

File: memprof/memtrace/Makefile [hunk 1]
```diff
@@ -73,17 +73,22 @@ ocamlopt :=  $(BEST_OCAMLOPT) $(flags)
...
 
 .PHONY: all allopt opt.opt # all, allopt and opt.opt are synonyms
 all: ocamlopt-memtrace.opt
 allopt: all
 opt.opt: allopt
 
-ocamlopt-memtrace.opt: $(native_modules)
+ocamlopt-memtrace.opt: $(ROOTDIR)/compilerlibs/ocamlcommon.cmxa \
+                       $(ROOTDIR)/compilerlibs/ocamloptcomp.cmxa \
+                       $(ROOTDIR)/driver/optmaindriver.cmx \
```
Comment (lthls) [inline memprof/memtrace/Makefile]: This line should be removed, `.depend` correctly tracks this dependency with your patch

File: Makefile [hunk 2]
```diff
@@ -734,7 +734,11 @@ ocamlopt.opt: compilerlibs/ocamlcommon.cmxa compilerlibs/ocamloptcomp.cmxa \
               $(OPTSTART:.cmo=.cmx)
 	$(CAMLOPT_CMD) $(LINKFLAGS) -o $@ $^
 
-ocamlopt-memtrace.opt: otherlibrariesopt
+ocamlopt-memtrace.opt: compilerlibs/ocamlcommon.cmxa \
+                       compilerlibs/ocamloptcomp.cmxa \
+                       driver/optmaindriver.cmx \
+                       otherlibrariesopt libraryopt \
+                       memprof/memtrace
```
Comment (lthls) [inline Makefile]: It seems unnecessary to have all those `cmx`/`cmxa` files here. I would put `ocamlopt.opt`, to specify that they should be kept in sync, and `otherlibrariesopt`, as it is an additional dependency of the memtrace version, and that's all.

**Comment by lthls** [comment]
I just noticed there's something wrong with the `depend` rule: it appends to `.depend` without removing previous entries. I think this should be fixed (and the cleaned up `.depend` commited).

File: memprof/memtrace/Makefile (hunk 1 above)
Comment (Gbury) [reply memprof/memtrace/Makefile]: done

File: Makefile (hunk 2 above)
Comment (Gbury) [reply Makefile]: done

**Comment by Gbury** [comment]
I fixed the depend append problem and cleaned the `.depend`.
This should be good to go.

**Pull request merged**